lib/web_ui/lib/src/engine/keyboard_binding.dart

```diff
@@ -0,0 +1,489 @@
+const int _kLogicalMetaRight = 0x0400000109;
+// Map logical keys for modifier keys to the functions that can get their
+// modifier flag out of an event.
+final Map<int, _ModifierGetter> _kLogicalKeyToModifierGetter = {
+  _kLogicalAltLeft: (FlutterHtmlKeyboardEvent event) => event.altKey,
+  _kLogicalAltRight: (FlutterHtmlKeyboardEvent event) => event.altKey,
+  _kLogicalControlLeft: (FlutterHtmlKeyboardEvent event) => event.ctrlKey,
+  _kLogicalControlRight: (FlutterHtmlKeyboardEvent event) => event.ctrlKey,
+  _kLogicalShiftLeft: (FlutterHtmlKeyboardEvent event) => event.shiftKey,
+  _kLogicalShiftRight: (FlutterHtmlKeyboardEvent event) => event.shiftKey,
+  _kLogicalMetaLeft: (FlutterHtmlKeyboardEvent event) => event.metaKey,
+  _kLogicalMetaRight: (FlutterHtmlKeyboardEvent event) => event.metaKey,
+};
+
+// After a keydown is received, this is the duration we wait for a repeat event
+// before we decide to synthesize a keyup event.
+//
+// On Linux and Windows, the typical ranges for keyboard repeat delay go up to
+// 1000ms. On Mac, the range goes up to 2000ms.
+const Duration _kKeydownCancelDurationNormal = Duration(milliseconds: 1000);
+const Duration _kKeydownCancelDurationMacOs = Duration(milliseconds: 2000);
+
+// ASCII for a, z, A, and Z
+const int _kCharLowerA = 0x61;
+const int _kCharLowerZ = 0x7a;
+const int _kCharUpperA = 0x41;
+const int _kCharUpperZ = 0x5a;
+bool isAlphabet(int charCode) {
+  return (charCode >= _kCharLowerA && charCode <= _kCharLowerZ)
+      || (charCode >= _kCharUpperA && charCode <= _kCharUpperZ);
+}
+
+const String _kPhysicalCapsLock = 'CapsLock';
+
+const String _kLogicalDead = 'Dead';
+
+const int _kWebKeyIdPlane = 0x00800000000;
+const int _kAutogeneratedMask = 0x10000000000;
+
+// Bits in a Flutter logical event to generate the logical key for dead keys.
+//
+// Logical keys for dead keys are generated by annotating physical keys with
+// modifiers (see `_getLogicalCode`).
+const int _kDeadKeyCtrl = 0x100000000000;
+const int _kDeadKeyShift = 0x200000000000;
+const int _kDeadKeyAlt = 0x400000000000;
+const int _kDeadKeyMeta = 0x800000000000;
+
+typedef DispatchKeyData = void Function(ui.KeyData data);
+
+/// Converts a floating number timestamp (in milliseconds) to a [Duration] by
+/// splitting it into two integer components: milliseconds + microseconds.
+Duration _eventTimeStampToDuration(num milliseconds) {
+  final int ms = milliseconds.toInt();
+  final int micro = ((milliseconds - ms) * Duration.microsecondsPerMillisecond).toInt();
+  return Duration(milliseconds: ms, microseconds: micro);
+}
+
+class KeyboardBinding {
+  /// The singleton instance of this object.
+  static KeyboardBinding? get instance => _instance;
+  static KeyboardBinding? _instance;
+
+  static void initInstance(html.Element glassPaneElement) {
+    if (_instance == null) {
+      _instance = KeyboardBinding._(glassPaneElement);
+      assert(() {
+        registerHotRestartListener(_instance!._reset);
```

Comment:
`_reset()` calls `_setup()`. I don't think you want to setup twice during a hot restart, right?

The callback passed to `registerHotRestartListener` should only do clean up. Setup will be executed later when the `KeyboardBinding` instance is re-created.